ResolvableFuture can only be called from within the same library group prefix

Execution failed for task ':workmanager:lint'.
> Lint found errors in the project; aborting build.

  Fix the issues identified by lint, or add the following to your build script to proceed with errors:
  ...
  android {
      lintOptions {
          abortOnError false
      }
  }
  ...
  Errors found:

  C:\src\flutter\.pub-cache\hosted\pub.dartlang.org\workmanager-0.5.0\android\src\main\kotlin\be\tramckrijte\workmanager\BackgroundWorker.kt:59: Error: ResolvableFuture can only be called from within the same library group prefix (referenced groupId=androidx.concurrent with prefix androidx from groupId=be.tramckrijte.workmanager) [RestrictedApi]
      private val resolvableFuture = ResolvableFuture.create<Result>()
                                     ~~~~~~~~~~~~~~~~
  C:\src\flutter\.pub-cache\hosted\pub.dartlang.org\workmanager-0.5.0\android\src\main\kotlin\be\tramckrijte\workmanager\BackgroundWorker.kt:132: Error: ResolvableFuture.set can only be called from within the same library group prefix (referenced groupId=androidx.concurrent with prefix androidx from groupId=be.tramckrijte.workmanager) [RestrictedApi]
              resolvableFuture.set(result)
                               ~~~



* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.

* Get more help at https://help.gradle.org

Deprecated Gradle features were used in this build, making it incompatible with Gradle 7.0.
Use '--warning-mode all' to show the individual deprecation warnings.
See https://docs.gradle.org/6.7/userguide/command_line_interface.html#sec:command_line_warnings

BUILD FAILED in 10m 33s
1057 actionable tasks: 143 executed, 914 up-to-date

--- a/.github/workflows/test.yml
+++ b/.github/workflows/test.yml
@@ -31,7 +31,7 @@ jobs:
       - name: Build iOS App
         run: cd example && flutter build ios --debug --no-codesign
       - name: Run native iOS tests
-        run: cd example/ios && xcodebuild -workspace Runner.xcworkspace -scheme Runner -sdk iphonesimulator -destination 'platform=iOS Simulator,name=iPhone 12,OS=15.2' test
+        run: cd example/ios && xcodebuild -workspace Runner.xcworkspace -scheme Runner -sdk iphonesimulator -destination 'platform=iOS Simulator,name=iPhone 12,OS=16.0' test
 
   native_android_tests:
     runs-on: macos-latest
@@ -90,6 +90,4 @@ jobs:
         with:
           api-level: ${{ matrix.api-level }}
           target: ${{ matrix.target }}
-          arch: x86_64
-          profile: Nexus 6
           script: cd example && flutter test integration_test/workmanager_integration_test.dart

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -1,3 +1,7 @@
+# next
+
+* Android: Migrate away from ResolvableFuture (#399).
+
 # 0.5.1
 
 * Call DartPluginRegistrant.ensureInitialized when isolate is starting

--- a/android/build.gradle
+++ b/android/build.gradle
@@ -2,7 +2,7 @@ group 'be.tramckrijte.workmanager'
 version '1.0-SNAPSHOT'
 
 buildscript {
-    ext.kotlin_version = '1.6.21'
+    ext.kotlin_version = '1.7.10'
     repositories {
         google()
         mavenCentral()
@@ -25,7 +25,7 @@ apply plugin: 'com.android.library'
 apply plugin: 'kotlin-android'
 
 android {
-    compileSdkVersion 31
+    compileSdkVersion 33
 
     sourceSets {
         main.java.srcDirs += 'src/main/kotlin'
@@ -46,5 +46,5 @@ dependencies {
     implementation "androidx.work:work-runtime:$work_version"
     implementation "androidx.concurrent:concurrent-futures:1.1.0"
 
-    testImplementation 'junit:junit:4.12'
+    testImplementation 'junit:junit:4.13.2'
 }

--- a/android/src/main/kotlin/be/tramckrijte/workmanager/BackgroundWorker.kt
+++ b/android/src/main/kotlin/be/tramckrijte/workmanager/BackgroundWorker.kt
@@ -5,7 +5,6 @@ import android.os.Handler
 import android.os.Looper
 import android.util.Log
 import androidx.concurrent.futures.CallbackToFutureAdapter
-import androidx.concurrent.futures.ResolvableFuture
 import androidx.work.ListenableWorker
 import androidx.work.WorkerParameters
 import com.google.common.util.concurrent.ListenableFuture
@@ -16,7 +15,6 @@ import io.flutter.embedding.engine.plugins.shim.ShimPluginRegistry
 import io.flutter.plugin.common.MethodCall
 import io.flutter.plugin.common.MethodChannel
 import io.flutter.view.FlutterCallbackInformation
-import kotlinx.coroutines.flow.callbackFlow
 import java.util.Random
 
 /***
@@ -61,7 +59,7 @@ class BackgroundWorker(
 
     private var completer: CallbackToFutureAdapter.Completer<Result>? = null
 
-    private var resolvableFuture = CallbackToFutureAdapter.getFuture<Result> { completer ->
+    private var resolvableFuture = CallbackToFutureAdapter.getFuture { completer ->
         this.completer = completer
         null
     }

--- a/android/src/main/kotlin/be/tramckrijte/workmanager/BackgroundWorker.kt
+++ b/android/src/main/kotlin/be/tramckrijte/workmanager/BackgroundWorker.kt
@@ -4,6 +4,7 @@ import android.content.Context
 import android.os.Handler
 import android.os.Looper
 import android.util.Log
+import androidx.concurrent.futures.CallbackToFutureAdapter
 import androidx.concurrent.futures.ResolvableFuture
 import androidx.work.ListenableWorker
 import androidx.work.WorkerParameters
@@ -15,6 +16,7 @@ import io.flutter.embedding.engine.plugins.shim.ShimPluginRegistry
 import io.flutter.plugin.common.MethodCall
 import io.flutter.plugin.common.MethodChannel
 import io.flutter.view.FlutterCallbackInformation
+import kotlinx.coroutines.flow.callbackFlow
 import java.util.Random
 
 /***
@@ -56,7 +58,13 @@ class BackgroundWorker(
     private var engine: FlutterEngine? = null
 
     private var startTime: Long = 0
-    private val resolvableFuture = ResolvableFuture.create<Result>()
+
+    private var completer: CallbackToFutureAdapter.Completer<Result>? = null
+
+    private var resolvableFuture = CallbackToFutureAdapter.getFuture<Result> { completer ->
+        this.completer = completer
+        null
+    }
 
     override fun startWork(): ListenableFuture<Result> {
         startTime = System.currentTimeMillis()
@@ -129,7 +137,7 @@ class BackgroundWorker(
         // No result indicates we were signalled to stop by WorkManager.  The result is already
         // STOPPED, so no need to resolve another one.
         if (result != null) {
-            resolvableFuture.set(result)
+            this.completer?.set(result)
         }
 
         // If stopEngine is called from `onStopped`, it may not be from the main thread.

--- a/example/android/app/build.gradle
+++ b/example/android/app/build.gradle
@@ -26,7 +26,7 @@ apply plugin: 'kotlin-android'
 apply from: "$flutterRoot/packages/flutter_tools/gradle/flutter.gradle"
 
 android {
-    compileSdkVersion 31
+    compileSdkVersion 33
 
     sourceSets {
         main.java.srcDirs += 'src/main/kotlin'
@@ -39,7 +39,7 @@ android {
     defaultConfig {
         applicationId "dev.fluttercommunity.workmanager.example"
         minSdkVersion 16
-        targetSdkVersion 31
+        targetSdkVersion 33
         versionCode flutterVersionCode.toInteger()
         versionName flutterVersionName
         testInstrumentationRunner "android.support.test.runner.AndroidJUnitRunner"
@@ -60,7 +60,7 @@ flutter {
 
 dependencies {
     implementation "org.jetbrains.kotlin:kotlin-stdlib-jdk7:$kotlin_version"
-    testImplementation 'junit:junit:4.12'
+    testImplementation 'junit:junit:4.13.2'
     androidTestImplementation 'com.android.support.test:runner:1.0.2'
     androidTestImplementation 'com.android.support.test.espresso:espresso-core:3.0.2'
 }

--- a/example/android/build.gradle
+++ b/example/android/build.gradle
@@ -1,5 +1,5 @@
 buildscript {
-    ext.kotlin_version = '1.6.21'
+    ext.kotlin_version = '1.7.10'
     repositories {
         google()
         mavenCentral()

